protocol/decoupled-hardfork-features.md

```diff
@@ -0,0 +1,308 @@
+complete. It effectively alone delayed the hardfork by a month or so.
+
+This could have been avoided by being able to implement the feature using
+a feature-specific toggle in the `RollupConfig`. The feature toggle would
+be toggled on in the [`superchain-registry`][scr] once the feature is ready.
+In this way, features could be implemented and left in client software
+without being connected to hardfork. The key part: *no changes would
+need to be made to client software to enable the feature*, only enabling
+the feature inside the [`superchain-registry`].
+
+
+# Problem Statement
+
+Bundling features within hardforks limits flexibility:
+
+- All features in a hardfork activate at once, preventing selective enablement
+  for testing or debugging.
+- Custom chains or testnets cannot easily remap or delay individual features
+  without forking entire client implementations.
+- CLI overrides are limited to hardfork timestamps, not per-feature adjustments.
+- The set of features for a given hardfork must be determined prior to
+  implementation, locking and committing to scope.
+
+This rigidity hinders iteration, increases deployment risks, and complicates
+superchain governance.
+
+
+# Current Approach
+
+- Hardforks are defined by named timestamps in the `RollupConfig`
+  (e.g., `fjordTime: uint64`).
+- Node logic checks `timestamp >= hardforkTime` to enable the all features
+  in the hardfork.
+- Features are implicitly tied to their hardfork; no per-feature granularity.
+- CLI overrides exist for hardfork timestamps (e.g., `--override.fjord=123456`
+  in `op-node`).
+    - Overrides in the `op-node` exist [here][op-overrides]. In the `kona-node`,
+      [here][kona-overrides].
+
+
+# Proposed Solution
+
+Decouple features from hardforks by modifying `RollupConfig`s to include a
+mapping from feature names to hardfork names (e.g., a `map[Feature]Fork`
+where the key is feature like "operatorFee" and value is a hardfork like "isthmus").
+If a feature is absent from the map, it remains inactive.
+
+### Key Changes
+
+- **Superchain Registry Updates:** Add a mapping from features to their hardfork
+  in the `superchain.toml` (superchain-wide config) as well as individual chain
+  configs if an override is needed.
+- **RollupConfig Update:** Add a field to map feature to its hardfork. For example,
+  a `features` field in the `RollupConfig` is a map from feature to hardfork name.
+- **Activation Check:** Add methods like `IsFeatureActive(timestamp uint64) bool`, which:
+    - Look up the hardfork for the feature.
+    - Retrieves the hardfork's timestamp (e.g., via existing fields like `IsthmusTime`).
+    - Returns true if `timestamp >= hardforkTimestamp`.
+- **Client-Level Overrides:** Extend CLI flags for direct per-feature timestamp overrides
+  (e.g., `--override.feature.eip4844=isthmus` or `--override.feature.eip4844.timestamp=123456`),
+  similar to current hardfork overrides.
```

Comment:
Do we actually need to do both of these? Seems like having it _either_ in the rollup config _or_ as a CLI flag would be sufficient and reduce the amount of code needing to be written for each feature.

Even if we create a framework that does make it easy to do both (and we should make it easy regardless), it seems like having two places to specify this would add confusion without really giving any extra flexibility. Would be interested if there's real use cases where you couldn't just edit the rollup config to change what's enabled.

Reply:
Good point. I will note that this is an optional change for posterity. I only included to be cohesive about how features could be configured.